src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs

```diff
@@ -718,5 +719,93 @@ private async ValueTask<T> CatchAndRethrowExceptionAsync<T>(Func<Task<T>> taskFu
                 throw exception;
             }
         }
+
+        public async ValueTask<IReadOnlyDictionary<Tuple<NuGetFramework, string>, IReadOnlyList<IPackageReferenceContextInfo>>> GetTransitivePackageOriginAsync(PackageIdentity transitivePackage, string projectId, CancellationToken ct)
+        {
+            /** Pseudocode
+            1. Get project restore graph 
+
+            2. Filter by packages
+
+            3. Foreach direct dependency d:
+                do DFS to look for transitive dependency
+
+                if found:
+                  Add to list
+
+            4. Return list
+            */
+            ct.ThrowIfCancellationRequested();
+
+            var singleProjectId = new[] { projectId };
+            var projectsList = await GetProjectsAsync(singleProjectId, ct);
+            var project = projectsList.FirstOrDefault();
+            var packageOrigins = new Dictionary<Tuple<NuGetFramework, string>, IReadOnlyList<IPackageReferenceContextInfo>>();
+
+            if (project != default && project is PackageReferenceProject prProject)
+            {
+                IList<ProjectModel.LockFileTarget> fxGraphList = await prProject.GetFullRestoreGraphAsync(ct);
+
+                if (fxGraphList != null)
+                {
+                    var pkgs = await GetInstalledAndTransitivePackagesAsync(singleProjectId, ct);
+
+                    var visited = new HashSet<object>();
+
+                    foreach (var targetFxGraph in fxGraphList)
+                    {
+                        var key = Tuple.Create(targetFxGraph.TargetFramework, targetFxGraph.RuntimeIdentifier);
+                        var list = new List<IPackageReferenceContextInfo>();
+
+                        foreach (var directPkg in pkgs.InstalledPackages) // are InstalledPackages direct dependencies only? Yes!
+                        {
+                            visited.Clear();
```

Comment:
If I understand this correctly, we try to find all the transitive dependencies of a top-level package. I am thinking, there may be a case where two top-level packages have same transitive dependency graph. If that is the case, clearing `visited` nodes may cause some redundant work. For e.g. Let's say a project target framework has a top-level dependency on `packageA` and `packageG` with following graph.

packageA --> packageB ---> packageC --> packageD
                |
                --->packageE --> packageF

packageG --> packageC --> packageD
 
Instead of maintaining HashSet for visited nodes, if we can maintain a dictionary of (visited packageId, bool result), then we may exit the DFS processing early. trade-off between time & space complexity.

Reply:
Great idea! Added.